### ISSUE TITLE
Zero copy

### ISSUE THREAD

**Issue by jfallows** [reaktivity/ignition#13]
Each nukleus in the pipeline receives frames over shared memory via a ring buffer data structure. `DATA` frames include payload by value so that payload is both read from and then also written back into a ring buffer by each nukleus until it reaches `tcp` where it is written to the network.

This enhancement aims to eliminate the requirement to copy the data payload into a ring buffer at each stage in the pipeline.

`BEGIN`, `DATA`, `ABORT`, `END`, `RESET` and `WINDOW` frames are simplified to `BEGIN`, `TRANSFER` and `ACK` frames with flags.

`BEGIN` `=>` `BEGIN`
`BEGIN` `END` `=>` `BEGIN(fin)`
`BEGIN` `DATA` `END` `=>` `BEGIN` `TRANSFER(fin)`
`BEGIN` `DATA` `ABORT` `=>` `BEGIN` `TRANSFER(rst)`
`WINDOW` => `ACK` (initial window now implicit)
`RESET` => `ACK(rst)`

The nukleus responsible for the source of data entering the pipeline should instead own the buffer slot associated with that stream. For example, when `tcp` establishes a new connection, it allocates a buffer slot and uses it to read data from the network, and after an initial `BEGIN` frame, it then sends a `TRANSFER` frame over shared memory with a reference to the buffer region containing data read from the network, instead of a copy.

The receiver of the `TRANSFER` frame at the next stage of the pipeline may need to decode the data, so it discovers the framing length and then delivers a decoded `TRANSFER` frame referring to only the decoded payload buffer sub-region. At this point the nukleus may choose to proactively send an `ACK` back to the source, indicating that the buffer sub-region containing the framing is no longer needed.

The receiver of the `TRANSFER` frame at the next stage of the pipeline may need to encode the data, so the encoded `TRANSFER` frame must contain a reference to multiple buffer regions, such as one for the framing with corresponding payload length, followed by same buffer region as that received for the data payload, both listed in one `TRANSFER` frame sent on down the pipeline.

The receiver of the `TRANSFER` frame at the next stage of the pipeline may need to write data to the network. For example, `tcp` may perform a gathering write of the list of received buffer regions as a single operation on the network socket. After the buffered data has been written to the network, the `tcp` nukleus sends an `ACK` frame back to the sender to notify that the buffer region(s) are no longer needed.

The receiver of the `ACK` frame at each of the previous stages of the pipeline will propagate the `ACK` frame back towards the source, such as `tcp` nukleus. This allows the acknowledged buffer region to be reused by `tcp` to read more data from the network.

Note that this approach no longer requires an initial `WINDOW` frame before sending data after the `BEGIN`, and back pressure is applied based on the rate of `ACK` frames received by the source of the data entering the pipeline and bounding the amount of buffer in use for each stream. In addition, a `BEGIN(fin)` or `BEGIN` `TRANSFER(fin)` can now be explicitly rejected by `ACK(rst)`. This implies that the reaktor cannot clean up the throttle handler after sending `BEGIN(fin)` or `TRANSFER(fin)`. Instead throttle handler clean up occurs when `ACK(fin)` (normal completion) or `ACK(rst)` (error completion) is received.

For single sender single receiver pipelines, the back pressure works by aligning the sender rate with receiver rate. In the case of single sender multiple receiver pipelines, where receivers may have different rates, then the sender is throttled by the rate of the slowest receiver. In cases where this is not desirable, it may be necessary to copy the data from the sender into an independent buffer slot to allow the timing of `ACK`s back to the sender to vary from the timing of the `ACK`s coming from the slowest receiver.

Each nukleus will need the ability to request a buffer slot, for `tcp` this is used to read data from the network socket, for other nuklei it may be necessary to create a distinct area for _encoded_ data, such as _encrypted_ data in `tls` nukleus. When an encoding buffer is used, then an `ACK` of the encoding buffer region should result in a transitive `ACK` of the original buffer. Chaining these `ACK`s together allows for proper end-to-end flow control on the pipeline.

We can make the guarantee to protocol decoders that partial protocol frames are provided in contiguous `TRANSFER` `regions`. We can also make the guarantee that protocol encoders will provide complete protocol frame headers. This bounds the amount of tracking information retained when partially decoding, as successive contiguous `TRANSFER` frames can be coalesced and treated as if received as a single `TRANSFER` frame.

There is an edge case at the end of the sender buffer when bytes are then accumulated at the beginning of the (circular) sender buffer, and the receiver may not be able to process a complete frame over the contiguous region. In this case, the receiver can view the non-contiguous regions as if contiguous, either by a cheap copy, or by accessing the data appropriately across the non-contiguous regions.

In the general case, the references to buffer regions will need to safely span threads as they flow from nukleus to nukleus through the pipeline. This involves taking care not to trigger false sharing of different buffer slots because they happen to both extend into the same cache line.

**Comment by cmebarrow** [comment]
@jfallows If there are multiple encoding nuklei on the pipeline will they each have to acquire their own buffer slot for each stream they are writing to? That would imply the use of multiple buffer slots per outgoing stream. Or would they be able to use different regions of a single buffer slot for multiple streams? In the latter case (or even in the former) it might be worth having a general API to help nuklei manage to reserve and free ranges within a slot.

**Comment by jfallows** [comment]
@cmebarrow Agreed, it makes sense to have the `reaktor` provide easy access to the buffer sizes needed by nuklei rather than have each such nukleus approach their requirement independently.

**Comment by jitsni** [comment]
* Since sub-regions in a block can be ACKed, source implementations (for e.g tcp) needs to keep track of the sub-regions before releasing an acquired block.
* Are ACK frames go to throttle buffers or stream buffers ? I think BEGIN, WRITE, ACK can all go to one buffer, then there won't be any throttle handler (so no need for cleaning up of throttle handler except ACKing regions).

**Comment by jfallows** [comment]
@jitsni: note that `ACK` is different from `release`.

**Comment by jfallows** [comment]
Renamed `WRITE` to `TRANSFER`.

**Comment by jfallows** [comment]
Added section about bounding the amount of tracking information that must be retained by protocol decoders for partial protocol frames.

**Comment by cmebarrow** [comment]
@jfallows  I have some questions about the "We can make the guarantee"  paragraph
- Should decoders reset (ACK(Rst)) if the rule about at most one  discontinuity is broken? Or maybe it's better to handle an indefinite number and treat the limit of two  as an optimization?  
- How to decode using flyweights across a discontinuity? Mem copy into a coalesced buffer?
- How can  a nukleus know  the maximum amount of data it may receive (i.e. the slot size  used by the source nukleus, e.g. tcp)? Kafka nukleus will need to know this in order to limit the  maximum fetch bytes and  avoid falling into the same bug  that I just fixed.

**Comment by jfallows** [comment]
@cmebarrow discontinuity is permitted if framing is complete, so decoders should `ACK(rst)` if the regions are non-contiguous _and_ framing is incomplete - the purpose is to reduce the amount of tracking information required to just a single high water mark to maintain awareness of decoding progress.

If the entire frame is formed of non-contiguous regions, then it is still feasible for the frame header to be isolated in one of those regions, so it is possible to decode these without copying. That is an optimization so copying into a reusable buffer to decode is a reasonable first step, before we fine tune to remove that copy for the common cases.

There is no longer a concept of window, so advance notice of how much data the receiver can handle, however this was previously a function of buffer size at the terminal receiver nukleus anyway. Given that the a reference to the sender buffer is now passed through the zerocopy pipeline instead, the terminal receiver nukleus is no longer bounded in the same way. Instead, the sender nukleus' ability to transfer more data is bound by the rate of acknowledgment from the receiver nukleus.

**Comment by jfallows** [comment]
Added overview of `TRANSFER(blk)`, `UNBLOCK` and `NACK`.

**Comment by jfallows** [comment]
Removed `TRANSFER(blk)`, `UNBLOCK` and `NACK`, and added overview of handling non-contiguous regions at wrap point in circular sender buffer.

**Comment by jfallows** [comment]
After prototyping this approach, decided to keep the original `WINDOW` based approach to provide a better bound on resource utilization and eliminate the complexity that is introduced when attempting to share references across nuklei.
